src/emacs-lisp-error-reporter.cpp

```diff
@@ -0,0 +1,92 @@
+// Copyright (C) 2020  Matthew "strager" Glazar
+// See end of file for extended copyright information.
+
+#include <ostream>
+#include <quick-lint-js/char8.h>
+#include <quick-lint-js/emacs-lisp-error-reporter.h>
+#include <quick-lint-js/emacs-location.h>
+#include <quick-lint-js/location.h>
+#include <quick-lint-js/optional.h>
+#include <quick-lint-js/padded-string.h>
+#include <quick-lint-js/token.h>
+
+namespace quick_lint_js {
+emacs_lisp_error_reporter::emacs_lisp_error_reporter(std::ostream &output)
+    : output_(output) {
+  this->output_ << "(";
+}
+
+void emacs_lisp_error_reporter::finish() { this->output_ << ")"; }
+
+void emacs_lisp_error_reporter::set_source(padded_string_view input) {
+  this->locator_.emplace(input);
+}
+
+#define QLJS_ERROR_TYPE(name, code, struct_body, format_call) \
+  void emacs_lisp_error_reporter::report(name e) {            \
+    format_error(e, this->format(code));                      \
+  }
+QLJS_X_ERROR_TYPES
+#undef QLJS_ERROR_TYPE
+
+emacs_lisp_error_formatter emacs_lisp_error_reporter::format(const char *code) {
+  QLJS_ASSERT(this->locator_.has_value());
+  return emacs_lisp_error_formatter(/*output=*/this->output_,
+                                    /*locator=*/*this->locator_,
+                                    /*code=*/code);
+}
+
+emacs_lisp_error_formatter::emacs_lisp_error_formatter(std::ostream &output,
+                                                       emacs_locator &locator,
+                                                       const char *code)
+    : output_(output), locator_(locator), code_(code) {}
+
+void emacs_lisp_error_formatter::write_before_message(
+    severity sev, const source_code_span &origin) {
+  std::string_view severity_type{};
+  if (sev == severity::note) {
+    return;
+  }
+  emacs_source_range r = this->locator_.range(origin);
+  // Emacs point starts at 1
+  emacs_source_position::offset_type beg = r.begin().offset + 1;
+  emacs_source_position::offset_type end = r.end().offset + 1;
+  this->output_ << "((" << beg << " . " << end << ") " << static_cast<int>(sev)
+                << " "
+                << "\"" << this->code_ << "\" \"";
+}
+
+void emacs_lisp_error_formatter::write_message_part(severity sev,
+                                                    string8_view message) {
+  if (sev == severity::note) {
+    return;
+  }
+  this->output_ << out_string8(message);
```

Comment:
Almost-blocking: What if the message contains `"`? We should escape `"` and possibly other characters, right?